spikeextractors/extractors/neuroscopeextractors/neuroscopeextractors.py

```diff
@@ -0,0 +1,451 @@
+        
+        shank_res_ids = [x[-1] for x in onlyfiles if x[-5:-2] == 'res' and x[-10:-6] != 'temp']
+        shank_clu_ids = [x[-1] for x in onlyfiles if x[-5:-2] == 'clu' and x[-10:-6] != 'temp']
+
+        assert shank_res_ids==shank_clu_ids, 'Unmatched .clu.%i and .res.%i files detected!'
+        
+        all_shanks_list_se = []
+        for shank_id in shank_res_ids:
+            resfile_path = '{}/{}.res.{}'.format(folder_path,SORTING_NAME,shank_id)
+            clufile_path = '{}/{}.clu.{}'.format(folder_path,SORTING_NAME,shank_id)
+    
+            all_shanks_list_se.append(NeuroscopeSortingExtractor(resfile_path=resfile_path,
+                                                                 clufile_path=clufile_path,
+                                                                 keep_mua_units=keep_mua_units))
+                    
+
+        MultiSortingExtractor.__init__(self,sortings=all_shanks_list_se)
+            
+        self._kwargs = {'folder_path': str(Path(folder_path).absolute()),
+                        'keep_mua_units': keep_mua_units,
+                        'exclude_shanks': exclude_shanks}
+    
+    
+    @staticmethod
+    def write_sorting(multisorting, save_path):
+        _, SORTING_NAME = os.path.split(save_path)
+        
+        counter = 1
+        save_xml = "{}/{}.xml".format(save_path,SORTING_NAME)
+        for sorting in multisorting.sortings:
+            # Create and save .res.%i and .clu.%i files from the current sorting object
+            save_res = "{}/{}.res.{}".format(save_path,SORTING_NAME,counter)
+            save_clu = "{}/{}.clu.{}".format(save_path,SORTING_NAME,counter)
+            counter += 1
+            unit_ids = sorting.get_unit_ids()
+            if len(unit_ids) > 0:
+                spiketrains = [sorting.get_unit_spike_train(u) for u in unit_ids]
+                res = np.concatenate(spiketrains).ravel()
+                clu = np.concatenate([np.repeat(i+1,len(st)) for i,st in enumerate(spiketrains)]).ravel() # i here counts from 0
+                res_sort = np.argsort(res)
+                res = res[res_sort]
+                clu = clu[res_sort]
+            else:
+                res = []
+                clu = []
+            
+            unique_ids = np.unique(clu)
+            n_clu = len(unique_ids)
+                
+            clu = np.insert(clu, 0, n_clu) # The +1 is necessary here b/c the convention for the base sorting object is from 1,...,nUnits
+    
+            np.savetxt(save_res, res, fmt='%i')
+            np.savetxt(save_clu, clu, fmt='%i')
+            
+        # create parameters file if none exists
+        if not os.path.isfile(save_xml):
+            soup = BeautifulSoup("",'xml')
+
+            new_tag = soup.new_tag('samplingrate')
+            new_tag.string = str(sorting.get_sampling_frequency())
+            soup.append(new_tag)
+
+            # write parameters file
+            f = open(save_xml, "w")
```

Comment:
generally, for files I would use context to make sure that the file is properly closed (even if you forget the close())

```
with open(save_xml, "w") as f:
    f.write(str(soup))
```

Reply:
Fixed in all cases! That is indeed good practice, I'll try to get into that habit.